feat: help for core flags

rishabh3112 opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. -->

**What kind of change does this PR introduce?**
Feature

**Did you add tests for your changes?**
No, will need

**If relevant, did you update the documentation?**
No

**Summary**
Adds help of flags as we have help for command from #1098.
<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
<!-- Try to link to an open issue for more information. -->

**Does this PR introduce a breaking change?**
No
<!-- If this PR introduces a breaking change, please describe the impact and a migration path for existing applications. -->

**Other information**
#1098 https://github.com/webpack/webpack-cli/issues/1053

jamesgeorge007 commented on lib/groups/help.js:
```diff
@@ -3,13 +3,13 @@ const chalk = require('chalk');
+    outputHelp(subject = null, isCommand = true) {
+        if (subject !== null) {
+            const info = isCommand ? require('../utils/commands').info : require('../utils/core-flags').info;
+            const options = isCommand ? info.find(cmd => cmd.name === subject || cmd.alias === subject) : info.find(flg => flg.name === subject.slice(2) || flg.alias === subject.slice(1));
```
This one seems a bit hard to read. Maybe use an if/else statement instead.

jamesgeorge007 commented on lib/groups/help.js:
```diff
@@ -3,13 +3,13 @@ const chalk = require('chalk');
+    outputHelp(subject = null, isCommand = true) {
+        if (subject !== null) {
+            const info = isCommand ? require('../utils/commands').info : require('../utils/core-flags').info;
+            const options = isCommand ? info.find(cmd => cmd.name === subject || cmd.alias === subject) : info.find(flg => flg.name === subject.slice(2) || flg.alias === subject.slice(1));
```
This one seems a bit hard to read. Maybe use an if/else statement instead.

rishabh3112 replied:
`if/else` would make `options` mutable. Instead used arrow function so as to not lose immutability.

jamesgeorge007 replied:
Yeah, makes sense :+1:

jamesgeorge007 commented on lib/utils/core-flags.js:
```diff
@@ -0,0 +1,14 @@
+    .reduce((arr, val) => arr.concat(val), []);
+
+module.exports.names = names;
+module.exports.info = core;
```
```js
module.exports = {names, core};
```

rishabh3112 replied:
I prefer current ones as I want to keep `info` and variable name instead of `core`.

pranshuchittora replied:
```

module.exports = {
    names,
    info:core
}

```

rishabh3112 replied:
Cool. It's night here, will update tommorow 👍

jamesgeorge007 replied:
We're all from the same time zone :sweat_smile:

jamesgeorge007 commented on lib/webpack-cli.js:
```diff
@@ -162,11 +162,17 @@ class WebpackCLI extends GroupHelper {
+            })[0] ||
+            flagNames.filter(name => {
+                return args.includes(name);
+            })[0];
```
Maybe there would be a potential scope for refactoring.

jamesgeorge007 replied:
Also, why not make use of the inline approach?

rishabh3112 replied:
@jamesgeorge007 refactored in latest commit.

jamesgeorge007 replied:
Much better :+1:

jamesgeorge007 commented on lib/groups/help.js:
```diff
@@ -3,13 +3,18 @@ const chalk = require('chalk');
+                if (isCommand) {
+                    return info.find(cmd => cmd.name === subject || cmd.alias === subject);
+                }
+                return info.find(flg => flg.name === subject.slice(2) || flg.alias === subject.slice(1));
```
```suggestion
                return info.find(flag => flag.name === subject.slice(2) || flag.alias === subject.slice(1));
```

rishabh3112 replied:
Made this change, also updated `cmd` to `command` too.

rishabh3112 commented:
@jamesgeorge007 awesome quick reviews 🐱‍👤👍

jamesgeorge007 commented on lib/groups/help.js:
```diff
@@ -3,13 +3,18 @@ const chalk = require('chalk');
-            const commands = require('../utils/commands').info;
+    outputHelp(subject = null, isCommand = true) {
+        if (subject !== null) {
+            const info = isCommand ? require('../utils/commands').info : require('../utils/core-flags').info;
```
https://github.com/webpack/webpack-cli/blob/3d6c24b8e345e7db48a5970d778f29168d5d3ba6/lib/utils/commands.js#L12

https://github.com/webpack/webpack-cli/blob/3d6c24b8e345e7db48a5970d778f29168d5d3ba6/lib/groups/help.js#L8

#1122

jamesgeorge007 commented on lib/webpack-cli.js:
```diff
@@ -162,11 +162,14 @@ class WebpackCLI extends GroupHelper {
+        const allNames = [...commandNames, ...flagNames];
+        const subject = allNames.filter(name => {
             return args.includes(name);
         })[0];
```
```js
        const subject = allNames.filter(name => args.includes(name))[0];
```
Maybe :thinking:

pranshuchittora replied:
A little bit difficult to understand at first glance ...

rishabh3112 commented:
## ℹ Status 
- All CI errors related to this PR has been resolved.
- All expected features implemented

ematipico commented on lib/utils/cli-flags.js:
```diff
@@ -182,27 +192,31 @@ module.exports = {
         },
         {
             name: 'sourcemap',
+            usage: '--sourcemap <sourcemap>',
```
Maybe we should show the alternatives provided by webpack:
`--sourcemap <sourcemap|eval|>` etc.

Basically what we have here: https://webpack.js.org/configuration/devtool/#devtool

ematipico commented on lib/utils/cli-flags.js:
```diff
@@ -182,27 +192,31 @@ module.exports = {
         {
             name: 'target',
             alias: 't',
+            usage: '--target <build target>',
```
Maybe we should show the targets available: https://webpack.js.org/configuration/target/#target

ematipico commented on lib/utils/cli-flags.js:
```diff
@@ -129,50 +131,58 @@ module.exports = {
         },
         {
             name: 'global',
+            usage: '--global',
```
The usage should be different:
`--global myVar ./global.js`

ematipico commented on lib/groups/help.js:
```diff
@@ -3,13 +3,18 @@ const chalk = require('chalk');
-    outputHelp(command = null) {
-        if (command !== null) {
-            const commands = require('../utils/commands').info;
+    outputHelp(subject = null, isCommand = true) {
```
You should revert the properties. Usually the ones that are not needed (the ones that are `undefined` or `null`) should go after the one that have mandatory values.

Also, are you use you want to have a `null`? I mean, if you don't pass anything `isCommand` will be `undefined` which is good for this case, right?

ematipico commented on lib/groups/help.js:
```diff
@@ -3,13 +3,18 @@ const chalk = require('chalk');
+    outputHelp(subject = null, isCommand = true) {
+        if (subject !== null) {
+            const info = isCommand ? require('../utils/commands').info : require('../utils/core-flags').info;
+            const options = (() => {
```
Why this is and inline IIFE? I don't think we need it.

```js
const options = info.find(commandOrFlag => {
   if (isCommand) {
     // return 
   }
   // return
})
```

rishabh3112 replied:
> Why this is

`options` is the `object` with details about `subject`.

> inline IIFE?

I was recommended to maintain immutability at each step and I didn't had any other implementation in my mind. Thus Opted for inline iffe. Will update it as per your suggestion in next commit.

> I don't think we need it.

Is there any other way to do this?

ematipico replied:
You can use `.reduce` if you really want to keep a functional approach :)

ematipico commented on lib/utils/core-flags.js:
```diff
@@ -0,0 +1,13 @@
+const { core } = require('./cli-flags');
+
+const names = core
```
Maybe also a test would help. It's up to you! We should improve the readability/usage of the code :)

rishabh3112 commented:
I updated the help and added documentation links for some of those.
![image](https://user-images.githubusercontent.com/15812317/72613091-a1a1e800-3954-11ea-8246-4cf7fdf825e0.png)

ematipico commented:
@rishabh3112 if you resolved the conflicts, I'd say we're ready to merge it!

rishabh3112 commented:
@ematipico rebased, and we are ready to go I think.
